pkg/cloud-controller-manager/loadBalancer.go

```diff
@@ -0,0 +1,326 @@
+	ctllbv1 "github.com/harvester/harvester-load-balancer/pkg/generated/controllers/loadbalancer.harvesterhci.io/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog/v2"
+)
+
+const (
+	defaultWaitIPTimeout = time.Second * 5
+)
+
+type LoadBalancerManager struct {
+	lbClient    ctllbv1.LoadBalancerClient
+	namespace   string
+	clusterName string
+}
+
+func newLoadBalancerManager(cfg *rest.Config, namespace string, clusterName string) (cloudprovider.LoadBalancer, error) {
+
+	lbFactory, err := ctllb.NewFactoryFromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &LoadBalancerManager{
+		lbClient:    lbFactory.Loadbalancer().V1alpha1().LoadBalancer(),
+		namespace:   namespace,
+		clusterName: clusterName,
+	}, nil
+}
+
+func (l *LoadBalancerManager) GetLoadBalancer(ctx context.Context, clusterName string, service *v1.Service) (status *v1.LoadBalancerStatus, exists bool, err error) {
+	name := l.GetLoadBalancerName(ctx, clusterName, service)
+	lb, err := l.lbClient.Get(l.namespace, name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil, false, nil
+		} else {
+			return nil, false, err
+		}
+	}
+
+	var ip string
+	if lb.Spec.Type == lbv1.External {
+		ip = lb.Status.ExternalAddress
+	} else {
+		ip = lb.Status.InternalAddress
+	}
+
+	return &v1.LoadBalancerStatus{
+		Ingress: []v1.LoadBalancerIngress{
+			{
+				IP: ip,
+			},
+		},
+	}, true, nil
+}
+
+func (l *LoadBalancerManager) GetLoadBalancerName(ctx context.Context, clusterName string, service *v1.Service) string {
+	// If users don't specify cluster name by the command flag, we adopt the cluster name from the cloud config file
+	if clusterName == "" || clusterName == "kubernetes" {
+		clusterName = l.clusterName
+	}
+	name := clusterName + "-" + service.Namespace + "-" + service.Name
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names
+	// The name contains no more than 253 characters
+	if len(name) > 253 {
+		name = name[:253]
+	}
+
+	return name
+}
+
+// EnsureLoadBalancer is to create/update a Harvester load balancer for the service and return the loadBalancerStatus with an IP
+// 1. watch the LB to get an IP asynchronously
+// 2. create or update lb
+// 3. wait for an ip and return the LoadBalancerStatus
+func (l *LoadBalancerManager) EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error) {
+	name := l.GetLoadBalancerName(ctx, clusterName, service)
+	ipChan := make(chan string)
+
+	lb, getErr := l.lbClient.Get(l.namespace, name, metav1.GetOptions{})
+	if getErr != nil && !errors.IsNotFound(getErr) {
+		return nil, getErr
+	}
+
+	// watch the lb to get an ip
+	w, err := l.lbClient.Watch(l.namespace, metav1.ListOptions{})
```

Comment:
Should we watch a specific lb filtered by FieldSelector

Reply:
We can add a label like `loadbalancer.harvesterhci.io/name` in the CR loadBalancer.

Reply:
TODO: Add a label in harvester-load-balancer